scripts/test_with_stripe_mock.go

```diff
@@ -0,0 +1,172 @@
+//
+//     go run test_with_stripe_mock.go
+//     go run test_with_stripe_mock.go ./charge
+//
+// The reason that we need a separate script for this is because Go's testing
+// infrastructure doesn't provide any kind of global hook that we can use to do
+// this work in only one place before any tests are run.
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+	"time"
+)
+
+const (
+	defaultStripeMockPort = "12112"
+
+	pathSpec     = "./testing/openapi/spec3.json"
+	pathFixtures = "./testing/openapi/fixtures3.json"
+)
+
+var portMatch = regexp.MustCompile(` port: (\d+)`)
+
+func main() {
+	var port string
+	var stripeMockProcess *os.Process
+
+	//
+	// Maybe start stripe-mock
+	//
+
+	if _, err := os.Stat(pathSpec); os.IsNotExist(err) {
+		port = os.Getenv("STRIPE_MOCK_PORT")
+		if port == "" {
+			port = defaultStripeMockPort
+		}
+		fmt.Printf("No custom spec file found, assuming stripe-mock is already running on port %v\n", port)
+	} else {
+		var err error
+		port, stripeMockProcess, err = startStripeMock()
+		if err != nil {
+			exitWithError(err)
+		}
+	}
+
+	//
+	// Run tests
+	//
+
+	err := runTests(port)
+	if err != nil {
+		stopStripeMock(stripeMockProcess)
+		exitWithError(err)
+	}
+
+	//
+	// Stop stripe-mock
+	//
+
+	// Try to cleanly stop stripe-mock, but in case we don't, it'll die anyway
+	// because it's executing as a subprocess.
+	stopStripeMock(stripeMockProcess)
+}
+
+//
+// Private functions
+//
+
+func exitWithError(err error) {
+	fmt.Fprintf(os.Stderr, "%v", err)
+	os.Exit(1)
+}
+
+func runTests(port string) error {
+	args := append([]string{"test"}, os.Args[1:]...)
+
+	// Defaults to `./...`, but also allows a specific package (or other CLI
+	// flags like `-test.v`) to be passed.
+	if len(args) == 1 {
+		args = append(args, "./...")
+	}
+
+	cmd := exec.Command("go", args...)
+
+	// Inherit this script's environment so that it's still possible to pass
+	// the test package flags like `GOCACHE=off`.
+	cmd.Env = append(
+		os.Environ(),
+		"STRIPE_MOCK_PORT="+port,
+	)
+
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stdout
+
+	err := cmd.Run()
+	if err != nil {
+		return fmt.Errorf("Error running tests: %v", err)
+	}
+
+	return nil
+}
+
+func startStripeMock() (string, *os.Process, error) {
+	fmt.Printf("Starting stripe-mock...\n")
+
+	cmd := exec.Command(
+		"stripe-mock",
+		"-https-port", "0", // stripe-mock will select a port
+		"-spec", pathSpec,
+		"-fixtures", pathFixtures,
+	)
+
+	cmd.Stderr = os.Stderr
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return "", nil, fmt.Errorf("Error starting stripe-mock: %v", err)
+	}
+
+	err = cmd.Start()
+	if err != nil {
+		return "", nil, fmt.Errorf("Error starting stripe-mock: %v", err)
+	}
+
+	b := make([]byte, 1024)
+	var port string
+
+	// We store the entire captured output because the string we're looking for
+	// may have appeared across a read boundary.
+	var stdoutBuffer bytes.Buffer
+
+	for {
+		n, err := stdout.Read(b)
```

Comment:
Just curious, wouldn't it be simpler to use [`ReadString`](https://golang.org/pkg/bytes/#Buffer.ReadString) with `\n` as the delimiter and check each line for the port?

Reply:
Good question! These IO constructs can all get a little confusing.

So the answer is that `stdout` here is not in fact a `bytes.Buffer`. `cmd.StdoutPipe` above returns an [`io.ReadCloser`](https://godoc.org/io#ReadCloser), which is a combination of a `Closer` and a `Reader`. [`Reader`](https://godoc.org/io#Reader) is an extremely simple interface that only provides us with one function: `Read`.

I think it's all stripped down this way because in this case Go has few guarantees around how that stdout pipe can be expected to behave. If you had a `ReadBytes`-type function on it, what would happen if the program never emitted another `\n`? It could either block forever, or you'd need to add a timeout mechanism.

Reply:
Makes sense! Thanks for the thorough answer.